From: RockOS <rockos@localhost>
Subject: [PATCH] chromium ffmpeg: fix x86 shift asm for newer binutils

Fix the bundled Chromium FFmpeg x86 inline assembly so QtWebEngine
5.15.14 builds with newer binutils/toolchains.

---
--- a/src/3rdparty/chromium/third_party/ffmpeg/libavcodec/x86/mathops.h
+++ b/src/3rdparty/chromium/third_party/ffmpeg/libavcodec/x86/mathops.h
@@ -112,20 +112,28 @@
 
 // avoid +32 for shift optimization (gcc should do that ...)
 #define NEG_SSR32 NEG_SSR32
-static inline  int32_t NEG_SSR32( int32_t a, int8_t s){
-    __asm__ ("sarl %1, %0\n\t"
-         : "+r" (a)
-         : "ic" ((uint8_t)(-s))
-    );
+static inline int32_t NEG_SSR32(int32_t a, int8_t s){
+    if (__builtin_constant_p(s))
+        __asm__ ("sarl %1, %0\n\t"
+                 : "+r" (a)
+                 : "i" (-s & 0x1F));
+    else
+        __asm__ ("sarl %1, %0\n\t"
+                 : "+r" (a)
+                 : "c" ((uint8_t)(-s)));
     return a;
 }
 
 #define NEG_USR32 NEG_USR32
 static inline uint32_t NEG_USR32(uint32_t a, int8_t s){
-    __asm__ ("shrl %1, %0\n\t"
-         : "+r" (a)
-         : "ic" ((uint8_t)(-s))
-    );
+    if (__builtin_constant_p(s))
+        __asm__ ("shrl %1, %0\n\t"
+                 : "+r" (a)
+                 : "i" (-s & 0x1F));
+    else
+        __asm__ ("shrl %1, %0\n\t"
+                 : "+r" (a)
+                 : "c" ((uint8_t)(-s)));
     return a;
 }
 
